This exercise will require you to pull some data from the Quadndl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [1]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case


Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/ ).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [2]:
# First, import the relevant modules
# I've imported pandas to test dataframes for fun. I sovle all the tasks with both lists and dataframes
import requests
import json
import collections
import pandas as pd

In [3]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
a = requests.get("https://www.quandl.com/api/v3/datasets/WIKI/FB.json")

In [4]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure


These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

***Task 1*** Collect Data

In [5]:
res = requests.get("https://www.quandl.com/api/v3/datasets/FSE/AFX_X/data.json?start_date=2017-01-01&end_date=2017-12-31&api_key=")

**Task 2 Convert the returned JSON oject into a Python dictionary**

In [6]:
data = res.json()

In [7]:
# I've created a list columns for future use as indexes
columns = data['dataset_data']['column_names']
print(columns)

['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover']


**Note:** I subsetted the data into a more organized list of lists (not sure if this was the easiest solution???)

In [8]:
L_data = data['dataset_data']['data']

***Task 3: Calculate what the highest and lowest opening prices were for the stock in this period.***

In [9]:
open_price = [item[1] for item in L_data if item[1]!=None]

In [10]:

print("The lowest open price was:", min(open_price), ".The highest opening price was:", max(open_price))

The lowest open price was: 34.0 .The highest opening price was: 53.11


**Task 4: What was the largest change in any one day (based on High and Low price)?**

In [11]:
change_high_low =[]
for item in L_data:
    high = item[columns.index('High')]
    low = item[columns.index('Low')]
    change_per_day = high - low
    change_high_low.append(change_per_day)

print("The largest change in one day is:", max(change_high_low))

The largest change in one day is: 2.8100000000000023


***Task 5: What was the largest change between any two days (based on Closing Price)?***


In [12]:
columns.index("Close")

4

In [13]:
L_change_closing = [item[4] for item in L_data]
max(L_change_closing) - min(L_change_closing)

19.03

The largest change between any two days based on closing price was $19.03

***Task 6: What was the average daily trading volume during this year?***

In [14]:
columns.index("Traded Volume")

6

In [15]:
L_trading_volume = [item[6] for item in L_data if item[6]!=None]
average = sum(L_trading_volume)/len(L_trading_volume)
print(average)

89124.33725490196


The average daily trading volume was 89,124

***Task 7: What was the median trading volume during this year? (Note: you may need to implement your own function for calculating the median.)***

In [16]:
def median(list):
    s = sorted(list)
    n = len(s)
    if n%2==0:
        return (s[n/2] + s[(n/2)-1])/2
    else: 
        return s[n//2]

median(L_trading_volume)

76286.0

The median trading volume is 76,286

# Using Dataframes: Easier Solutions for Fun


In [17]:
columns = data['dataset_data']["column_names"]
d = data['dataset_data']['data']
df = pd.DataFrame(data=d, columns=columns)
df

Date   Open   High    Low  Close  Change  Traded Volume   Turnover  \
0    2017-12-29  51.76  51.94  51.45  51.76     NaN        34640.0  1792304.0   
1    2017-12-28  51.65  51.82  51.43  51.60     NaN        40660.0  2099024.0   
2    2017-12-27  51.45  51.89  50.76  51.82     NaN        57452.0  2957018.0   
3    2017-12-22  51.05  51.50  50.92  51.32     NaN        71165.0  3641949.0   
4    2017-12-21  51.16  51.52  50.90  51.40     NaN       120649.0  6179433.0   
..          ...    ...    ...    ...    ...     ...            ...        ...   
250  2017-01-06  34.91  35.21  34.91  35.04     NaN        27507.0   964046.0   
251  2017-01-05  35.02  35.20  34.73  35.06     NaN        48412.0  1692326.0   
252  2017-01-04  35.48  35.51  34.75  35.19     NaN        54408.0  1906810.0   
253  2017-01-03  35.90  35.93  35.34  35.48     NaN        70618.0  2515473.0   
254  2017-01-02  34.99  35.94  34.99  35.80     NaN        44700.0  1590561.0   

    Last Price of the Day Daily Traded Units Daily Turnover  
0                    None               None           None  
1                    None               None           None  
2                    None               None           None  
3                    None               None           None  
4                    None               None           None  
..                    ...                ...            ...  
250                  None               None           None  
251                  None               None           None  
252                  None               None           None  
253                  None               None           None  
254                  None               None           None  

[255 rows x 11 columns]

**Task 3: Calculate what the highest and lowest opening prices were for the stock in this period**

In [18]:
df.Open.max(), df.Open.min()

(53.11, 34.0)

The highest opening price was 53.11 and the lowest opening price was 34.00

***Task 4 What was the largest change in any one day (based on High and Low price)?***

In [19]:
df_change = df['High']-df['Low']
max(df_change)

2.8100000000000023

The largest change in one day was $2.81

***Task 5: What was the largest change between any two days (based on Closing Price)?***


In [20]:
df.Close.max() - df.Close.min()

19.03

***Task 6: What was the average daily trading volume during this year?***

In [21]:
df["Traded Volume"].mean()

89124.33725490196

***Task 7: What was the median trading volume during this year? (Note: you may need to implement your own function for calculating the median.)***

In [22]:
df["Traded Volume"].median()

76286.0